Unity-Technologies/ScriptableRenderPipeline
# Fix for: "Properties from the Blackboard can be dragged into the Graph when Blackboard is disabled"

**Summary:**
Fix for https://fogbugz.unity3d.com/f/cases/1187828/

**Update:**
Also fixes this issue as well: https://fogbugz.unity3d.com/f/cases/1211132/

Before:
![NoMoreInvisibleBlackboardProperties1](https://user-images.githubusercontent.com/3393041/69125087-da5b3900-0a59-11ea-9f90-5563e3f1db69.gif)

After:
![NoMoreInvisibleBlackboardProperties0](https://user-images.githubusercontent.com/3393041/69125089-dcbd9300-0a59-11ea-8dac-c7b039bd55f6.gif)

Implementation Notes:
* Because uielements.visible could sometimes result in elements still being selectable / draggable, these toggle-able windows are now removed & readded to the view upon toggle.
* ^ This is the same as how VFX works.
* Helper method ensures that the Main Preview & Blackboard are always in the same order (and any new sub windows), they will appear in the reverse order that they appear in ToggleSubWindows(). Fyi using Insert() on something that has already been inserted will simply move it and using RemoveFromHierarchy() on something that has already been removed does nothing, which is why ToggleSubWindows() doesn't bother checking for its current status.

QA Notes:
* This bug is super finicky, and really is a workaround for a bug in Graph View that may never get fixed. Sometimes it was other parts of the blackboard that is select-able (which you will be able to debug via UI Elements Debugger).

Manually tested:
* All combinations of toggle.
* Resizing the graph while the elements are removed, then readding them: location still makes sense.
* That the view order of the sub windows is always consistent with what it was before.
* Using the UI Elements Debugger to confirm removal and order.

---

**Yamato**: (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline/tree/sg%252Ffix%252F1187828/.yamato%252Fupm-ci-shadergraph.yml%2523All_ShaderGraph_trunk
(Master compare: https://yamato.prd.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline/tree/master/.yamato%252Fupm-ci-packages.yml%2523all_package_ci/1009336/job)
